pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ConsumerHandler.java

```diff
@@ -276,8 +284,62 @@ private void handleAck(ConsumerCommand command) throws IOException {
         // We should have received an ack
         MessageId msgId = MessageId.fromByteArrayWithTopic(Base64.getDecoder().decode(command.messageId),
                 topic.toString());
-        consumer.acknowledgeAsync(msgId).thenAccept(consumer -> numMsgsAcked.increment());
-        checkResumeReceive();
+        if ("cumulative".equals(command.ackType)) {
+            if (allowCumulativeAck) {
+                // If not pull mode, then we need to calculate how many messages have been acked in a cumulative ack request
+                // to properly calculate how many more messages we should dispatch to consumer. Use precise backlog size before
+                // and after the cumulative ack for the calculation.
+                // Use a writelock to prevent individual ack during the process which might cause incorrect result.
+                if (!this.pullMode) {
+                    try {
+                        pendingMessagesLock.writeLock().lock();
+                        MessageIdImpl messageId = (MessageIdImpl) msgId;
+                        long messagesToBeAcked = service.getAdminClient().topics().getNumberOfUnackedMessages(topic.toString(), subscription, messageId.getLedgerId(), messageId.getEntryId());
```

Comment:
yeah, we can skip checking and directly start receiving if no backlog as there won't be pending ack message.
but in other case we still need to check if pending ack message is below threshold to resume deliver
so checking messagesToBeAcked instead of backlog here.